[M2E] Cannot run Bnd OSGi X Launcher(s) from shortcuts menu

Ubuntu version `21.04`
Eclipse version `2021-09 (4.21.0)`
Bndtools versions tested: `6.1.0.DEV-202110112327-g86c7e1e`, `6.0.0.REL-202110041835-gc077e57`

When I try, I just get an exception in the logs.

```
!ENTRY org.eclipse.debug.ui 4 120 2021-10-13 12:22:32.503
!MESSAGE Error logged from Debug UI: 
!STACK 1
org.eclipse.core.runtime.CoreException: Plug-in bndtools.m2e was unable to load class bndtools.launch.JUnitShortcut.
	at org.eclipse.core.internal.registry.osgi.RegistryStrategyOSGI.throwException(RegistryStrategyOSGI.java:212)
	at org.eclipse.core.internal.registry.osgi.RegistryStrategyOSGI.createExecutableExtension(RegistryStrategyOSGI.java:198)
	at org.eclipse.core.internal.registry.ExtensionRegistry.createExecutableExtension(ExtensionRegistry.java:920)
	at org.eclipse.core.internal.registry.ConfigurationElement.createExecutableExtension(ConfigurationElement.java:246)
	at org.eclipse.core.internal.registry.ConfigurationElementHandle.createExecutableExtension(ConfigurationElementHandle.java:63)
	at org.eclipse.debug.internal.ui.launchConfigurations.LaunchShortcutExtension.getDelegate(LaunchShortcutExtension.java:404)
	at org.eclipse.debug.internal.ui.launchConfigurations.LaunchShortcutExtension.launch(LaunchShortcutExtension.java:428)
	at org.eclipse.debug.internal.ui.actions.LaunchShortcutAction.runInternal(LaunchShortcutAction.java:88)
	at org.eclipse.debug.internal.ui.actions.LaunchShortcutAction.runWithEvent(LaunchShortcutAction.java:135)
	at org.eclipse.jface.action.ActionContributionItem.handleWidgetSelection(ActionContributionItem.java:580)
	at org.eclipse.jface.action.ActionContributionItem.lambda$4(ActionContributionItem.java:414)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:5884)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1522)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1548)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1531)
	at org.eclipse.swt.widgets.Widget.notifyListeners(Widget.java:1310)
	at org.eclipse.jface.action.ActionContributionItem.lambda$8(ActionContributionItem.java:1207)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:5884)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1522)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:5126)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4576)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1150)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1041)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:644)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:551)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:156)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:152)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:136)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:401)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:659)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:596)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1467)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1440)
Caused by: java.lang.ClassNotFoundException: bndtools.launch.JUnitShortcut cannot be found by bndtools.m2e_6.1.0.202110112327-SNAPSHOT
	at org.eclipse.osgi.internal.loader.BundleLoader.findClass(BundleLoader.java:506)
	at org.eclipse.osgi.internal.loader.ModuleClassLoader.loadClass(ModuleClassLoader.java:170)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:519)
	at org.eclipse.osgi.internal.framework.EquinoxBundle.loadClass(EquinoxBundle.java:616)
	at org.eclipse.core.internal.registry.osgi.RegistryStrategyOSGI.createExecutableExtension(RegistryStrategyOSGI.java:196)
	... 43 more
```

--- a/DEV_README.md
+++ b/DEV_README.md
@@ -214,7 +214,10 @@ will not be able to restart it.
    type of the extension that you are implementing.
    2. Ensure that the `component.name` property is the same as the `id`
    field in the `plugin.xml` of the plugin that registers the extension (set
-   in step 1.i). By default, this is the fqn of the implementation class.
+   in step 1.i). By default, this is the fqn of the implementation class. (Note:
+   there are some cases where you might need the `id` to be different from
+   the `component.name` - in such cases, see the section "Overriding the
+   `component.name`" below.)
    3. Set the correct component scope. Usually the component will need to be
    prototype scope, otherwise the implementation instances will interfere with
    each other - however, if your component implementation is stateless, then you
@@ -242,11 +245,23 @@ a new backing component arrives or is closed. `RunAction` uses this feature
 to enable/disable itself depending on whether or not the backing component is
 present.
 
+## Overriding the `component.name`
+
+By default, the `ExtensionFacade` will use the extension element's `id` attribute to
+look for a DS component with a `component.name` that matches. However, sometimes you
+want two extensions that point to the same DS component. Because having extensions with
+duplicate ids is not a good idea, the `ExtensionFacade` provides an alternative method
+to specify the `component.name` to match: by adding `:<component.name>` to the end
+of the class.
+
+If you need to combine the factory mode with this feature, leave the name of the class
+blank (eg, `class="org.bndtools.facade.ExtensionFacade::my.component.name"`).
+
 ## Examples of the `ExtensionFacade` in action
 
 The prototype example of how to use the `ExtensionFacade` is in the 
 `org.bndtools.launch` bundle, which contains the launch-related code
 for Bndtools. This bundle was used as a proving ground for the
 `ExtensionFacade`'s initial development. The extensions are registered in
-`bndtools.core`'s `plugin.xml`, using the `ExtensionFacade`
-configured as described above.
+`bndtools.core` and `bndtools.m2e`'s `plugin.xml` files, using the
+`ExtensionFacade` configured as described above.

--- a/bndtools.core/src/org/bndtools/facade/ExtensionFacade.java
+++ b/bndtools.core/src/org/bndtools/facade/ExtensionFacade.java
@@ -45,9 +45,9 @@ public class ExtensionFacade<T> implements IExecutableExtension, IExecutableExte
 	String						propertyName;
 	Class<T>					downstreamClass;
 	Object						data;
-	static final BundleContext	bc	= Optional.ofNullable(FrameworkUtil.getBundle(ExtensionFacade.class))
+	static final BundleContext	bc			= Optional.ofNullable(FrameworkUtil.getBundle(ExtensionFacade.class))
 		.map(Bundle::getBundleContext)
-		.orElseGet(null);
+		.orElse(null);
 
 	@Override
 	public Object create() throws CoreException {
@@ -190,37 +190,52 @@ public class ExtensionFacade<T> implements IExecutableExtension, IExecutableExte
 
 		consoleLog.debug("{} Initializing facade, propName: \"{}\", data: \"{}\"", this, propertyName, data);
 
-		if (data != null && !data.toString()
-			.isEmpty()) {
-			try {
-				String epId = config.getDeclaringExtension()
-					.getExtensionPointUniqueIdentifier();
+		if (data != null) {
+			final String dataString = data.toString();
 
-				IExtensionPoint ep = Platform.getExtensionRegistry()
-					.getExtensionPoint(epId);
-				String bp = ep.getContributor()
-					.getName();
+			final int index = dataString.indexOf(':');
 
-				Optional<Bundle> b = Stream.of(bc.getBundles())
-					.filter(x -> bp.equals(x.getSymbolicName()))
-					.findFirst();
-
-				if (b.isPresent()) {
-					consoleLog.debug("{} Attempting to load \"{}\" from bundle: {}", this, data, b.get());
-					@SuppressWarnings("unchecked")
-					final Class<T> clazz = (Class<T>) b.get()
-						.loadClass(data.toString());
-					downstreamClass = clazz;
-				} else {
-					consoleLog.debug("Using our classloader");
-					@SuppressWarnings("unchecked")
-					final Class<T> clazz = (Class<T>) Class.forName(data.toString());
-					downstreamClass = clazz;
+			String className;
+			if (index < 0) {
+				className = dataString;
+			} else {
+				className = dataString.substring(0, index);
+				if (index < dataString.length() - 1) {
+					this.id = dataString.substring(index + 1);
 				}
-			} catch (ClassNotFoundException e) {
-				consoleLog.error("{} exception:", this, e);
-				throw new CoreException(
-					new Status(IStatus.ERROR, getClass(), 0, "Downstream interface for " + id + " not found", e));
+			}
+
+			if (!className.isEmpty()) {
+				try {
+					String epId = config.getDeclaringExtension()
+						.getExtensionPointUniqueIdentifier();
+
+					IExtensionPoint ep = Platform.getExtensionRegistry()
+						.getExtensionPoint(epId);
+					String bp = ep.getContributor()
+						.getName();
+
+					Optional<Bundle> b = Stream.of(bc.getBundles())
+						.filter(x -> bp.equals(x.getSymbolicName()))
+						.findFirst();
+
+					if (b.isPresent()) {
+						consoleLog.debug("{} Attempting to load \"{}\" from bundle: {}", this, data, b.get());
+						@SuppressWarnings("unchecked")
+						final Class<T> clazz = (Class<T>) b.get()
+							.loadClass(className);
+						downstreamClass = clazz;
+					} else {
+						consoleLog.debug("Using our classloader");
+						@SuppressWarnings("unchecked")
+						final Class<T> clazz = (Class<T>) Class.forName(className);
+						downstreamClass = clazz;
+					}
+				} catch (ClassNotFoundException e) {
+					consoleLog.error("{} exception:", this, e);
+					throw new CoreException(
+						new Status(IStatus.ERROR, getClass(), 0, "Downstream interface for " + id + " not found", e));
+				}
 			}
 		}
 		try {

--- a/bndtools.m2e/_plugin.xml
+++ b/bndtools.m2e/_plugin.xml
@@ -35,7 +35,7 @@
 	<extension
 			point="org.eclipse.debug.ui.launchShortcuts">
 		<shortcut
-				class="bndtools.launch.JUnitShortcut"
+				class="org.bndtools.facade.ExtensionFacade:org.eclipse.debug.ui.ILaunchShortcut2:bndtools.launch.JUnitShortcut"
 				icon="icons/bricks_junit.png"
 				id="bndtools.m2e.launch.junitShortcut"
 				label="Bnd OSGi Test Launcher (JUnit)"
@@ -70,7 +70,7 @@
 		</shortcut>
 
 		<shortcut
-				class="bndtools.launch.RunShortcut"
+				class="org.bndtools.facade.ExtensionFacade:org.eclipse.debug.ui.ILaunchShortcut2:bndtools.launch.RunShortcut"
 				icon="icons/bricks_run.png"
 				id="bndtools.m2e.launch.runShortcut"
 				label="Bnd OSGi Run Launcher"

